specs/PushNotifications/ToastNotifications-spec.md

```diff
@@ -0,0 +1,1026 @@
+    auto toastHistory = ToastNotificationManager::Default().History();
+    co_await toastHistory.RemoveWithTagGroupAsync(tag, group);
+}
+```
+
+## Toast Progress Updates
+
+Sometimes a developer would like to show progress bar related updates in a toast:
+
+![Toast Progress](https://docs.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/images/toast-progressbar-annotated.png)
+
+To accomplish that, the developer will need to use the ToastProgressData construct.
+
+```cpp
+// Send first Toast Progress Update
+void SendUpdatableToastWithProgress()
+{
+    winrt::hstring tag = L"weekly-playlist";
+    winrt::hstring group = L"downloads";
+
+    winrt::hstring payload = L"<toast launch=\"action = viewDownload &amp; downloadId = 9438108\">"
+                                L"<visual>"
+                                    L"<binding template = \"ToastGeneric\">"
+                                        L"<text>Downloading this week's new music...</text>"
+                                        L"<progress"
+                                            L"title = \"{progressTitle}\""
+                                            L"value = \"{progressValue}\""
+                                            L"valueStringOverride = \"{progressValueString}\""
+                                            L"status = \"{progressStatus}\" />"
+                                    L"</binding>"
+                                L"</visual>"
+                            L"</toast>";
+
+    XmlDocument doc;
+    doc.LoadXml(payload);
+    ToastNotification toast(doc);
+    toast.Tag(tag);
+    toast.Group(group);
+
+    // Assign initial values for first toast progress UI
+    ToastProgressData data;
+    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
+    data.Value(0.6); // Binds to {progressValue} in xml payload
+    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload
+    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload
+    data.SequenceNumber(1);
+
+    toast.ProgressData(data);
+    ToastNotificationManager::Default().ShowToast(toast);
+}
+
+// Send subsequent progress updates
+winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
+{
+    winrt::hstring tag = L"weekly-playlist";
+    winrt::hstring group = L"downloads";
+
+    // Assign new values
+    // Note that you only need to assign values that changed. In this example we don't assign progressStatus since we don't need to change it
+    ToastProgressData data;
+    data.Value(0.7); // Binds to {progressValue} in xml payload
+    data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload
+    data.SequenceNumber(2);
+
+    auto result = co_await ToastNotificationManager::Default().UpdateToastProgressDataAsync(data, tag, group);
+    if (result != ToastProgressResult::Succeeded)
+    {
+        LOG_HR_MSG(E_UNEXPECTED, "Toast Progress Update Failed!");
```

Comment:
it would be good if UpdateToastProgressDataAsync returned something more useful than succeeded / not succeeded. Can it be a struct with an hr? or some way to get back to the error context...

Reply:
We kind of mapped the behavior in WinAppSDK to mirror the Windows SDK return values for the Progress operation. So we would like to be consistent here since exposing internal HRESULT's may lead to a whole new can of worms.

    // The Result for a Toast Progress related operation
    enum ToastProgressResult
    {
        Succeeded, // The progress operation succeeded
        Failed, // The progress operation failed
        NotificationNotFound, // The progress operation failed to find a toast to process updates
    };

Reply:
DISCUSSION: Succeeded = was updated.
NotificationNotFound = valid call, but notification is gone.
Any other error = exception (app messed up).
We can RoOriginateError with useful diagnostic strings.